modules/core/04-channel/keeper/upgrade.go

```diff
@@ -86,17 +88,249 @@ func (k Keeper) WriteUpgradeInitChannel(
 	channelID string,
 	upgradeSequence uint64,
 	channelUpgrade types.Channel,
-) {
+) error {
 	defer telemetry.IncrCounter(1, "ibc", "channel", "upgrade-init")
 
+	channel, found := k.GetChannel(ctx, portID, channelID)
+	if !found {
+		return errorsmod.Wrapf(types.ErrChannelNotFound, "failed to retrieve channel %s on port %s", channelID, portID)
+	}
+
+	// assign directly the fields that are modifiable.
+	// counterparty fields may not be changed.
+	channel.Version = channelUpgrade.Version
+	channel.Ordering = channelUpgrade.Ordering
+	channel.ConnectionHops = channelUpgrade.ConnectionHops
+
+	// prevent silent failures if fields that are not allowed to be modified are changed in the upgrade channel
+	// request.
+	if !reflect.DeepEqual(channel, channelUpgrade) {
+		return errorsmod.Wrapf(types.ErrInvalidChannel, "expected: %v but got: %v", channel, channelUpgrade)
+	}
+
 	k.SetChannel(ctx, portID, channelID, channelUpgrade)
 	k.Logger(ctx).Info("channel state updated", "port-id", portID, "channel-id", channelID, "previous-state", types.OPEN.String(), "new-state", types.INITUPGRADE.String())
 
 	emitChannelUpgradeInitEvent(ctx, portID, channelID, upgradeSequence, channelUpgrade)
+	return nil
+}
+
+// ChanUpgradeTry is called by a module to accept the first step of a channel upgrade
+// handshake initiated by a module on another chain.
```

Comment:
Maybe it would be good to add something here in the good about the upgrade sequence that is returned, because if this function returns a not `nil` error, then `upgradeSequence` will be zero, but in fact the upgrade sequence stored could be different.